# Phase 3 — Data Cleaning

## Real Estate Investment Advisor

This notebook performs systematic data cleaning on the India housing prices dataset.

The raw dataset is preserved separately. Cleaning operations are applied to a working copy, and the cleaned dataset is saved to the processed data directory.

### Objectives

- Load the raw dataset
- Verify dataset structure
- Handle missing values
- Remove duplicate records if present
- Validate categorical values
- Validate numerical values
- Preserve important source information
- Document known structural inconsistencies
- Save the cleaned dataset

In [28]:
import pandas as pd
import numpy as np
import os

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [29]:
raw_path = "../data/raw/india_housing_prices.csv"

df = pd.read_csv(raw_path)

print("Dataset loaded successfully.")
print("Shape:", df.shape)

Dataset loaded successfully.
Shape: (250000, 23)


In [30]:
df.head()

,ID,State,City,Locality,Property_Type,BHK,Size_in_SqFt,Price_in_Lakhs,Price_per_SqFt,Year_Built,Furnished_Status,Floor_No,Total_Floors,Age_of_Property,Nearby_Schools,Nearby_Hospitals,Public_Transport_Accessibility,Parking_Space,Security,Amenities,Facing,Owner_Type,Availability_Status
0,1,Tamil Nadu,Chennai,Locality_84,Apartment,1,4740,489.76,0.10,1990,Furnished,22,1,35,10,3,High,No,No,"Playground, Gym, Garden, Pool, Clubhouse",West,Owner,Ready_to_Move
1,2,Maharashtra,Pune,Locality_490,Independent House,3,2364,195.52,0.08,2008,Unfurnished,21,20,17,8,1,Low,No,Yes,"Playground, Clubhouse, Pool, Gym, Garden",North,Builder,Under_Construction
2,3,Punjab,Ludhiana,Locality_167,Apartment,2,3642,183.79,0.05,1997,Semi-furnished,19,27,28,9,8,Low,Yes,No,"Clubhouse, Pool, Playground, Gym",South,Broker,Ready_to_Move
3,4,Rajasthan,Jodhpur,Locality_393,Independent House,2,2741,300.29,0.11,1991,Furnished,21,26,34,5,7,High,Yes,Yes,"Playground, Clubhouse, Gym, Pool, Garden",North,Builder,Ready_to_Move
4,5,Rajasthan,Jaipur,Locality_466,Villa,4,4823,182.90,0.04,2002,Semi-furnished,3,2,23,4,9,Low,No,Yes,"Playground, Garden, Gym, Pool, Clubhouse",East,Builder,Ready_to_Move


## 1. Dataset Structure

Before applying any cleaning operations, the structure and data types of the raw dataset are verified.

In [31]:
print("Shape:", df.shape)
print("\nData Types:")
print(df.dtypes)

Shape: (250000, 23)

Data Types:
ID                                  int64
State                                 str
City                                  str
Locality                              str
Property_Type                         str
BHK                                 int64
Size_in_SqFt                        int64
Price_in_Lakhs                    float64
Price_per_SqFt                    float64
Year_Built                          int64
Furnished_Status                      str
Floor_No                            int64
Total_Floors                        int64
Age_of_Property                     int64
Nearby_Schools                      int64
Nearby_Hospitals                    int64
Public_Transport_Accessibility        str
Parking_Space                         str
Security                              str
Amenities                             str
Facing                                str
Owner_Type                            str
Availability_Status                   str
d

## 2. Missing Value Check

The dataset is checked for missing values across all columns.

In [32]:
missing_summary = pd.DataFrame({
    "Missing_Count": df.isnull().sum(),
    "Missing_Percentage": (df.isnull().sum() / len(df)) * 100
})

missing_summary

,Missing_Count,Missing_Percentage
ID,0,0.0
State,0,0.0
City,0,0.0
Locality,0,0.0
Property_Type,0,0.0
BHK,0,0.0
Size_in_SqFt,0,0.0
Price_in_Lakhs,0,0.0
Price_per_SqFt,0,0.0
Year_Built,0,0.0


## 3. Duplicate Record Check

Duplicate rows are identified before cleaning.

In [33]:
duplicate_count = df.duplicated().sum()

print("Duplicate rows:", duplicate_count)

Duplicate rows: 0


In [34]:
if duplicate_count > 0:
    df = df.drop_duplicates().copy()
    print("Duplicate rows removed.")
else:
    print("No duplicate rows found. Dataset unchanged.")

No duplicate rows found. Dataset unchanged.


## 4. Unique ID Validation

The uniqueness of the property ID is verified.

In [35]:
duplicate_ids = df["ID"].duplicated().sum()

print("Duplicate IDs:", duplicate_ids)
print("Unique IDs:", df["ID"].nunique())
print("Total records:", len(df))

Duplicate IDs: 0
Unique IDs: 250000
Total records: 250000


In [36]:
if duplicate_ids == 0:
    print("All property IDs are unique.")
else:
    print("Duplicate property IDs require investigation.")

All property IDs are unique.


## 5. Categorical Text Standardization Check

Text-based categorical columns are inspected for leading/trailing spaces and inconsistent formatting.

In [37]:
categorical_columns = [
    "State",
    "City",
    "Locality",
    "Property_Type",
    "Furnished_Status",
    "Public_Transport_Accessibility",
    "Parking_Space",
    "Security",
    "Amenities",
    "Facing",
    "Owner_Type",
    "Availability_Status"
]

for col in categorical_columns:
    leading_trailing_spaces = (
        df[col].astype(str) != df[col].astype(str).str.strip()
    ).sum()

    print(f"{col}: {leading_trailing_spaces} records with leading/trailing spaces")

State: 0 records with leading/trailing spaces
City: 0 records with leading/trailing spaces
Locality: 0 records with leading/trailing spaces
Property_Type: 0 records with leading/trailing spaces
Furnished_Status: 0 records with leading/trailing spaces
Public_Transport_Accessibility: 0 records with leading/trailing spaces
Parking_Space: 0 records with leading/trailing spaces
Security: 0 records with leading/trailing spaces
Amenities: 0 records with leading/trailing spaces
Facing: 0 records with leading/trailing spaces
Owner_Type: 0 records with leading/trailing spaces
Availability_Status: 0 records with leading/trailing spaces


In [38]:
for col in categorical_columns:
    df[col] = df[col].astype(str).str.strip()

In [39]:
print("Text columns standardized successfully.")

Text columns standardized successfully.


## 6. Categorical Value Validation

Categorical columns are checked against the values observed during dataset inspection.

In [40]:
for col in categorical_columns:
    print(f"\n{col}")
    print(df[col].value_counts())


State
State
Odisha            12681
Tamil Nadu        12629
West Bengal       12622
Gujarat           12578
Delhi             12552
Telangana         12539
Maharashtra       12533
Punjab            12516
Uttar Pradesh     12508
Uttarakhand       12501
Assam             12496
Kerala            12487
Jharkhand         12480
Andhra Pradesh    12462
Chhattisgarh      12456
Madhya Pradesh    12451
Karnataka         12424
Rajasthan         12402
Bihar             12369
Haryana           12314
Name: count, dtype: int64

City
City
Coimbatore        6461
Ahmedabad         6411
Silchar           6404
Durgapur          6387
Cuttack           6358
Vijayawada        6344
Jamshedpur        6335
Bhubaneswar       6323
New Delhi         6306
Kochi             6305
Ludhiana          6295
Bhopal            6284
Hyderabad         6273
Noida             6271
Haridwar          6267
Warangal          6266
Bilaspur          6263
Dwarka            6246
Lucknow           6237
Kolkata           6235
Dehradun  

## 7. Numerical Feature Validation

Numerical columns are checked for invalid, negative, zero, or logically impossible values.

In [41]:
numerical_columns = [
    "ID",
    "BHK",
    "Size_in_SqFt",
    "Price_in_Lakhs",
    "Price_per_SqFt",
    "Year_Built",
    "Floor_No",
    "Total_Floors",
    "Age_of_Property",
    "Nearby_Schools",
    "Nearby_Hospitals"
]

print("Numerical columns:")
print(numerical_columns)

Numerical columns:
['ID', 'BHK', 'Size_in_SqFt', 'Price_in_Lakhs', 'Price_per_SqFt', 'Year_Built', 'Floor_No', 'Total_Floors', 'Age_of_Property', 'Nearby_Schools', 'Nearby_Hospitals']


In [42]:
validation_results = []

for col in numerical_columns:
    validation_results.append({
        "Column": col,
        "Negative_Count": (df[col] < 0).sum(),
        "Zero_Count": (df[col] == 0).sum(),
        "Minimum": df[col].min(),
        "Maximum": df[col].max()
    })

numerical_validation = pd.DataFrame(validation_results)

numerical_validation

,Column,Negative_Count,Zero_Count,Minimum,Maximum
0,ID,0,0,1.0,250000.00
1,BHK,0,0,1.0,5.00
2,Size_in_SqFt,0,0,500.0,5000.00
3,Price_in_Lakhs,0,0,10.0,500.00
4,Price_per_SqFt,0,2534,0.0,0.99
5,Year_Built,0,0,1990.0,2023.00
6,Floor_No,0,8203,0.0,30.00
7,Total_Floors,0,0,1.0,30.00
8,Age_of_Property,0,0,2.0,35.00
9,Nearby_Schools,0,0,1.0,10.00


## 8. Property-Level Logical Validation

Relationships between related numerical variables are checked for logical consistency.

In [43]:
print("BHK <= 0:", (df["BHK"] <= 0).sum())

print("Size <= 0:", (df["Size_in_SqFt"] <= 0).sum())

print("Price <= 0:", (df["Price_in_Lakhs"] <= 0).sum())

print("Total_Floors <= 0:", (df["Total_Floors"] <= 0).sum())

print("Floor_No < 0:", (df["Floor_No"] < 0).sum())

print("Age_of_Property < 0:", (df["Age_of_Property"] < 0).sum())

print("Year_Built outside 1900-2025:",
      ((df["Year_Built"] < 1900) | (df["Year_Built"] > 2025)).sum())

BHK <= 0: 0
Size <= 0: 0
Price <= 0: 0
Total_Floors <= 0: 0
Floor_No < 0: 0
Age_of_Property < 0: 0
Year_Built outside 1900-2025: 0


In [44]:
calculated_age = 2025 - df["Year_Built"]

age_difference = df["Age_of_Property"] - calculated_age

print("Age inconsistencies:", (age_difference != 0).sum())
print("Maximum absolute difference:", age_difference.abs().max())

Age inconsistencies: 0
Maximum absolute difference: 0


In [45]:
floor_inconsistency = df["Floor_No"] > df["Total_Floors"]

print("Floor inconsistencies:", floor_inconsistency.sum())
print(
    "Percentage:",
    round(floor_inconsistency.mean() * 100, 2),
    "%"
)

Floor inconsistencies: 116304
Percentage: 46.52 %


## 9. Price per SqFt Validation

The supplied `Price_per_SqFt` field is compared against a value calculated from the property's price and size.

In [46]:
df["Calculated_Price_per_SqFt"] = (
    df["Price_in_Lakhs"] * 100000
) / df["Size_in_SqFt"]

price_difference = (
    df["Price_per_SqFt"] -
    df["Calculated_Price_per_SqFt"]
)

print("Mean absolute difference:",
      price_difference.abs().mean())

print("Maximum absolute difference:",
      price_difference.abs().max())

print("Median calculated price per sqft:",
      df["Calculated_Price_per_SqFt"].median())

print("Median supplied price per sqft:",
      df["Price_per_SqFt"].median())

Mean absolute difference: 13058.151179072182
Maximum absolute difference: 99181.01
Median calculated price per sqft: 9244.747593959637
Median supplied price per sqft: 0.09


In [47]:
price_ratio = (
    df["Price_per_SqFt"] /
    df["Calculated_Price_per_SqFt"]
)

print("Records where supplied value differs substantially:",
      (price_ratio < 0.5).sum() + (price_ratio > 1.5).sum())

Records where supplied value differs substantially: 250000


In [48]:
df.drop(columns=["Calculated_Price_per_SqFt"], inplace=True)

print("Temporary investigation column removed.")
print("Current dataset shape:", df.shape)

Temporary investigation column removed.
Current dataset shape: (250000, 23)


## 10. Final Data Cleaning Decisions

The following cleaning decisions are applied:

1. No missing values were found.
2. No duplicate rows were found.
3. All property IDs are unique.
4. Text columns were stripped of leading/trailing whitespace.
5. Valid numerical values are retained.
6. `Age_of_Property` is internally consistent with `Year_Built`.
7. Floor inconsistencies are retained because correcting or removing 46.52% of the records would not be justified from the available data.
8. The original `Price_per_SqFt` column is retained in the cleaned dataset for traceability. A corrected price-per-square-foot feature will be created during Feature Engineering.
9. No statistical outliers are removed because the observed numerical ranges are plausible for the dataset and the IQR analysis did not identify meaningful outliers in the major numerical variables.

In [49]:
# Remove exact duplicate rows if any
df = df.drop_duplicates().copy()

# Remove leading/trailing whitespace from categorical columns
for col in categorical_columns:
    df[col] = df[col].astype(str).str.strip()

print("Basic cleaning completed.")
print("Shape:", df.shape)

Basic cleaning completed.
Shape: (250000, 23)


In [50]:
final_check = pd.DataFrame({
    "Missing_Values": df.isnull().sum(),
    "Unique_Values": df.nunique(),
    "Data_Type": df.dtypes.astype(str)
})

final_check

,Missing_Values,Unique_Values,Data_Type
ID,0,250000,int64
State,0,20,str
City,0,42,str
Locality,0,500,str
Property_Type,0,3,str
BHK,0,5,int64
Size_in_SqFt,0,4501,int64
Price_in_Lakhs,0,48700,float64
Price_per_SqFt,0,100,float64
Year_Built,0,34,int64


In [51]:
print("Rows:", len(df))
print("Columns:", len(df.columns))

print("\nDuplicate rows:", df.duplicated().sum())
print("Duplicate IDs:", df["ID"].duplicated().sum())

print("\nNegative BHK:", (df["BHK"] < 0).sum())
print("Invalid Size:", (df["Size_in_SqFt"] <= 0).sum())
print("Invalid Price:", (df["Price_in_Lakhs"] <= 0).sum())
print("Negative Age:", (df["Age_of_Property"] < 0).sum())

Rows: 250000
Columns: 23

Duplicate rows: 0
Duplicate IDs: 0

Negative BHK: 0
Invalid Size: 0
Invalid Price: 0
Negative Age: 0


In [52]:
processed_path = "../data/processed"

os.makedirs(processed_path, exist_ok=True)

cleaned_file = "../data/processed/india_housing_prices_cleaned.csv"

df.to_csv(cleaned_file, index=False)

print("Cleaned dataset saved successfully.")
print(cleaned_file)

Cleaned dataset saved successfully.
../data/processed/india_housing_prices_cleaned.csv


In [53]:
import os

print("File exists:", os.path.exists(cleaned_file))

if os.path.exists(cleaned_file):
    print(
        "File size:",
        round(os.path.getsize(cleaned_file) / (1024 * 1024), 2),
        "MB"
    )

File exists: True
File size: 39.45 MB


In [54]:
cleaned_df = pd.read_csv(cleaned_file)

print("Reloaded dataset shape:", cleaned_df.shape)
print("Columns:", len(cleaned_df.columns))
print("Missing values:", cleaned_df.isnull().sum().sum())
print("Duplicate rows:", cleaned_df.duplicated().sum())

Reloaded dataset shape: (250000, 23)
Columns: 23
Missing values: 0
Duplicate rows: 0


# Phase 3 — Data Cleaning Summary

The raw India housing dataset was successfully cleaned and validated.

### Final Dataset

- Records: 250,000
- Features: 23
- Missing values: 0
- Duplicate rows: 0
- Duplicate IDs: 0

### Cleaning Performed

- Verified dataset structure
- Checked missing values
- Checked duplicate records
- Verified unique property IDs
- Standardized categorical text formatting
- Validated numerical ranges
- Verified property age consistency
- Documented floor-number inconsistencies
- Investigated the supplied Price_per_SqFt field
- Preserved the original source values for traceability

### Known Dataset Limitations

Two important source-data issues were identified:

1. `Floor_No > Total_Floors` occurs in approximately 46.52% of records.
2. `Price_per_SqFt` does not correspond to the conventional calculation based on price and property size.

These fields are not artificially corrected during cleaning. The appropriate derived features will be addressed during Feature Engineering.

### Output

The cleaned dataset has been saved as:

`data/processed/india_housing_prices_cleaned.csv`